File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/address/AssetIdDecoder.java
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.android.address;

import java.util.Objects;

/** Parses canonical public asset identifiers. */
public final class AssetIdDecoder {

  private AssetIdDecoder() {}

  /** Parsed canonical public asset identifier. */
  public static final class AssetId {
    private final AssetDefinition definition;
    private final String accountId;
    private final Long dataspaceId;

    AssetId(AssetDefinition definition, String accountId, Long dataspaceId) {
      this.definition = Objects.requireNonNull(definition, "definition");
      this.accountId = Objects.requireNonNull(accountId, "accountId");
      this.dataspaceId = dataspaceId;
    }

    public AssetDefinition definition() {
      return definition;
    }

    public String accountId() {
      return accountId;
    }

    public Long dataspaceId() {
      return dataspaceId;
    }
  }

  /** Decoded asset definition carrying the canonical Base58 address. */
  public static final class AssetDefinition {
    private final String address;

    AssetDefinition(String address) {
      this.address = Objects.requireNonNull(address, "address");
    }

    public String address() {
      return address;
    }
  }

  /** Checks whether the given value is a canonical Base58 asset-definition id. */
  public static boolean isCanonical(String value) {
    if (value == null) {
      return false;
    }
    try {
      decode(value);
      return true;
    } catch (IllegalArgumentException ex) {
      return false;
    }
  }

  /** Parses a canonical Base58 asset-definition id. */
  public static AssetId decode(String assetId) {
    Objects.requireNonNull(assetId, "assetId");
    final String trimmed = assetId.trim();
    if (!trimmed.equals(assetId) || trimmed.isEmpty()) {
      throw new IllegalArgumentException(
          "AssetId must use canonical unprefixed Base58 asset-definition form");
    }
    AssetDefinitionIdEncoder.parseAddressBytes(trimmed);
    return new AssetId(new AssetDefinition(trimmed), "", null);
  }

  /**
   * Validates a canonical asset-definition address and returns it.
   */
  public static AssetDefinition decodeDefinition(String definitionId) {
    Objects.requireNonNull(definitionId, "definitionId");
    final String trimmed = definitionId.trim();
    if (!trimmed.equals(definitionId) || trimmed.isEmpty()) {
      throw new IllegalArgumentException(
          "Asset definition id must use canonical unprefixed Base58 form");
    }
    AssetDefinitionIdEncoder.parseAddressBytes(trimmed);
    return new AssetDefinition(trimmed);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/address/AssetIdDecoder.kt
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.sdk.address

/**
 * Parses canonical public asset identifiers.
 */
object AssetIdDecoder {

    class AssetId(
        @JvmField val definition: AssetDefinition,
        @JvmField val accountId: String,
        @JvmField val dataspaceId: Long?,
    )

    /**
     * Checks whether the given value is a canonical public asset literal.
     */
    @JvmStatic
    fun isCanonical(value: String?): Boolean {
        if (value == null) {
            return false
        }
        return try {
            decode(value)
            true
        } catch (_: IllegalArgumentException) {
            false
        }
    }

    /**
     * Parses `<asset-definition-id>#<i105-account-id>` with an optional `#dataspace:<id>` suffix.
     */
    @JvmStatic
    fun decode(assetId: String): AssetId {
        val trimmed = assetId.trim()
        require(trimmed == assetId && trimmed.isNotEmpty()) {
            "AssetId must use canonical public form"
        }

        val parts = trimmed.split('#')
        require(parts.size == 2 || parts.size == 3) {
            "AssetId must use '<asset-definition-id>#<i105-account-id>' with optional '#dataspace:<id>' suffix"
        }

        val definitionAddress = parts[0]
        AssetDefinitionIdEncoder.parseAddressBytes(definitionAddress)

        val parsedAccount = try {
            AccountAddress.parseEncodedIgnoringCurveSupport(parts[1], null).address
        } catch (ex: AccountAddressException) {
            throw IllegalArgumentException("AssetId.account must use canonical i105 form", ex)
        }
        val canonicalAccountId = try {
            parsedAccount.toI105(AccountAddress.DEFAULT_I105_DISCRIMINANT)
        } catch (ex: AccountAddressException) {
            throw IllegalArgumentException("AssetId.account must use canonical i105 form", ex)
        }

        val dataspaceId = when (parts.size) {
            2 -> null
            else -> parseDataspace(parts[2])
        }

        return AssetId(AssetDefinition(definitionAddress), canonicalAccountId, dataspaceId)
    }

    /**
     * Validates a canonical asset-definition address and returns it as an [AssetDefinition].
     */
    @JvmStatic
    fun decodeDefinition(definitionId: String): AssetDefinition {
        val trimmed = definitionId.trim()
        require(trimmed == definitionId && trimmed.isNotEmpty()) {
            "Asset definition id must use canonical unprefixed Base58 form"
        }
        AssetDefinitionIdEncoder.parseAddressBytes(trimmed)
        return AssetDefinition(trimmed)
    }

    private fun parseDataspace(scopeLiteral: String): Long {
        val match = Regex("^dataspace:(\\d+)$").matchEntire(scopeLiteral)
            ?: throw IllegalArgumentException(
                "AssetId.scope must use 'dataspace:<id>' when present"
            )
        return match.groupValues[1].toLong()
    }
}
